### PR TITLE
Added L0 tests

### PR DESCRIPTION
**Pull request by sundargs2000**

### CODE REVIEW

File: __tests__/main.test.ts [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+        }) 
+    }
+});
+
+describe('Testing all functions of class PipelineHelper', () => {
+    test('PipelineHelper.EnsureValidPipeline() - throw error if definition not found', () => {
+        expect(() => PipelineHelper.EnsureValidPipeline('someProject', 'somePipeline', [])).toThrow(new PipelineNotFoundError(`Pipeline named "${'somePipeline'}" not found in project "${'someProject'}"`));
+    });
+
+    test('PipelineHelper.EnsureValidPipeline() - throw error if more than one definition found', () => {
```
Comment (Josh-01) [inline __tests__/main.test.ts:68]: Remove 'PipelineHelper.' from all tests

File: __tests__/main.test.ts [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+    test('PipelineHelper.equals() - return if strings are equal', () => {
+        expect(PipelineHelper.equals(null, null)).toBeTruthy();
+        expect(PipelineHelper.equals('a', null)).toBeFalsy();
+        expect(PipelineHelper.equals(null, 'a')).toBeFalsy();
+        expect(PipelineHelper.equals('a', 'a ')).toBeTruthy();
+        expect(PipelineHelper.equals('a', 'A')).toBeTruthy();
+        expect(PipelineHelper.equals('a', 'b')).toBeFalsy();
+    });
+
+    test('PipelineHelper.processEnv() - return specified env variable varaiable', () => {
```
Comment (Josh-01) [inline __tests__/main.test.ts:81]: typo

File: __tests__/main.test.ts [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+        expect(PipelineHelper.getErrorAndWarningMessageFromBuildResult([
+            { message: 'FirstIgnoredMessage', result: 0},
+            { message: 'SecondIgnoredMessage', result: 1},
+        ])).toMatchObject({
+            errorMessage: '',
+            warningMessage: 'FirstIgnoredMessage,SecondIgnoredMessage',
+        });
+    });
+
+    test('PipelineHelper.getErrorAndWarningMessageFromBuildResult() - server errors also which comes not in form of array', () => {
```
Comment (Josh-01) [inline __tests__/main.test.ts:119]: what is being tested here ?

File: __tests__/main.test.ts [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+        expect(PipelineHelper.getErrorAndWarningMessageFromBuildResult(
+            { message: 'ErrorMessage' } as any
+        )).toMatchObject({
+            errorMessage: 'ErrorMessage',
+            warningMessage: '',
+        });
+    });
+
+    test('PipelineHelper.getErrorAndWarningMessageFromBuildResult() - server errors also which comes not in form of array', () => {
```
Comment (Josh-01) [inline __tests__/main.test.ts:128]: what is being tested here?

File: __tests__/main.test.ts [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+            { serverError: { message: 'ServerErrorMessage'  } } as any
+        )).toMatchObject({
+            errorMessage: 'ServerErrorMessage',
+            warningMessage: '',
+        });
+    });
+});
+
+describe('Testing all functions of class UrlParser', () => {
+    test('UrlParser.GetProjectName() - return project name from project URL', () => {
```
Comment (Josh-01) [inline __tests__/main.test.ts:139]: Remove 'UrlParser.' from all tests

File: __tests__/main.test.ts [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+        expect(() => UrlParser.GetCollectionUrlBase('')).toThrow('Project url is null or empty. Specify the valid project url and try again');
+    });
+
+    test('UrlParser.GetCollectionUrlBase() - throw error if invalid url', () => {
+        expect(() =>  UrlParser.GetCollectionUrlBase('/')).toThrow(`Failed to parse project url: "${'/'}". Specify the valid project url (eg, https://dev.azure.com/organization/project-name or https://server.example.com:8080/tfs/DefaultCollection/project-name)) and try again.`);
+    });
+});
+
+describe('Testing all functions of class PipelineRunner', () => {
+    test('start() - run a build using given parameters', async () => {
```
Comment (Josh-01) [inline __tests__/main.test.ts:168]: what is being tested here ?

File: __tests__/main.test.ts [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+            },
+            reason: 943,
+            sourceBranch: 'releases',
+            sourceVersion: 'sampleSha',
+        }
+        expect(mockQueueBuild).toBeCalledWith(expectedBuild, 'my-project', true);
+        expect(core.setFailed).toBeCalled();
+    });
+
+    test('start() - set core failed if results has errors', async () => {
```
Comment (Josh-01) [inline __tests__/main.test.ts:269]: make the test descriptions more meaningful/descriptive

**Review by Josh-01 — CHANGES_REQUESTED** [review]

File: __tests__/main.test.ts (hunk 4 above)
Comment (sundargs2000) [reply __tests__/main.test.ts:128]: As in? I was trying to capture the different if clauses to show how the `_getErrorMessageFromServer` functino parses the result object to return error messages.

File: __tests__/main.test.ts (hunk 6 above)
Comment (sundargs2000) [reply __tests__/main.test.ts:168]: go through a regular run to show what functions are called and cover that part of code.

File: __tests__/main.test.ts [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+    test('getErrorAndWarningMessageFromBuildResult() - server errors also which comes not in form of array', () => {
+        expect(PipelineHelper.getErrorAndWarningMessageFromBuildResult(
+            { message: 'ErrorMessage' } as any
+        )).toMatchObject({
+            errorMessage: 'ErrorMessage',
+            warningMessage: '',
+        });
+    });
+
+    test('getErrorAndWarningMessageFromBuildResult() - server errors also which comes not in form of array', () => {
```
Comment (Josh-01) [inline __tests__/main.test.ts:128]: this and previous error seem to be the same ? please fix this

File: __tests__/main.test.ts (hunk 4 above)
Comment (Josh-01) [reply __tests__/main.test.ts]: 2 tests have the same description - specify exactly what is tested in each in short

**Review by Josh-01 — APPROVED** [review]
:white_check_mark:

**Pull request merged**